src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -677,12 +786,42 @@ private struct PutCountersData
                         throw new InvalidDataException($"Remote Counter-Group document is missing '{DbIds}' property on '{documentId}'. Shouldn't happen");
                     }
 
+                    sourceData.TryGet(CounterNames, out BlittableJsonReaderArray sourceCounterNames);
                     BlittableJsonReaderObject data;
 
                     if (table.SeekOnePrimaryKeyPrefix(documentKeyPrefix, out _) == false)
                     {
                         // simplest case of having no counters for this document, can use the raw data from the source as-is
 
+                        if (sourceCounterNames == null)
+                        {
+                            // 4.2 source
+                            // need to create @names array and to lower the property names in 'sourceCounters'
+
+                            var originalNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                            var propDetails = new BlittableJsonReaderObject.PropertyDetails();
+                            sourceCounters.Modifications = new DynamicJsonValue(sourceCounters);
+
+                            for (int i = 0; i < sourceCounters.Count; i++)
```

Comment:
Here to, if this was an array, it would be easier to run though.
Note that we'll usually not trigger this work, so we want to avoid those costs if possible.